### PR TITLE
Add tests for Quoting

### PR DESCRIPTION
**Pull request by siefkenj**
Add minimal tests for `Quoting` found in `quoting.js`. I don't have a good sample of different emails to pull from, so the tests consist of one real email and several made up ones. More should be added if someone has sample emails.

These APIs are called in exactly one place in `messageIFrame.jsx`. All methods could be consolidated to a single method `wrapBlockquotes` or something like that and then called once instead of being called in a sequence in `messageIFrame.jsx`.

All internal-use-only methods of `Quoting` have been prefixed with `_`. I think there are better ways to handle the quoting code. Right now it passes around a DOM and mutates it, which is no good if data needs to be passed between a webextension and the Thunderbird API. It would be nice if this turned into a string-based API, though I don't know the performance implications.

Also, I'm not sure how to test the `outlook` quoting method. It seems to rely on `getComputedStyle` which is only available for DOM nodes that are actually inserted into a real page. I don't have any samples of these types of messages, so I cannot analyze to find another way to do it.

### CODE REVIEW

**Comment by siefkenj** [comment]
Upgraded Travis to `node-14` so that `?.` syntax would be recognized.

**Comment by Standard8** [comment]
Just to note I've seen this, I hope to get to it in the next day or so.

File: addon/tests/quoting.test.js [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+/* eslint-env jest */
+
+// Standard imports for all tests
+const { esmImport } = require("./utils");
+
+// Prettier is used to normalize the html formatting so we can reliably use it to compare HTML with
+// text diffing.
+const prettier = require("prettier");
```
Comment (Standard8) [inline addon/tests/quoting.test.js:11]: Nice idea.

**Review by Standard8 — CHANGES_REQUESTED** [review]
Sorry, I keep forgetting to actually finish this review off.

I think this is really good to have some tests - thank you. There's one comment inline to potentially address.

Re string manipulation vs DOM, I think I'd be concerned that we'd end up writing or including our own specialist parser, when the DOM would do it for us. Obviously there's expense in using the DOM as well. So I'm not sure there's an easy "best" solution either way, which makes me tempted to keep it as it is for now.

File: addon/tests/quoting.test.js [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+          <style type="text/css" style="display: none">
+            P {
+              margin-top: 0;
+              margin-bottom: 0;
+            }
+          </style>
+
+          <div class="moz-text-html" lang="x-western">
+            <div style="font-family: Calibri, Helvetica, sans-serif; font-size: 12pt; color: rgb(0, 0, 0)">This is really good to hear</div>
+            <div style="font-family: Calibri, Helvetica, sans-serif; font-size: 12pt; color: rgb(0, 0, 0)">
```
[the remark this answers is not in the document]
Comment (siefkenj) [reply addon/tests/quoting.test.js:31]: I can remove them. I left them in because this was from an actual email (the other examples were hand-crafted).

**Comment by siefkenj** [comment]
Thanks for the review :-).

As for string manipulation, I wasn't suggesting writing a DOM parser. I think `new DOMParser()` will work just fine for that. I was suggesting changing the interface to a string-based interface (it might parse to a DOM behind the scenes).

This would come with a slight performance penalty of parsing once to escape quoted blocks, and then TB parsing again when actually displaying the message. But, I don't think it will be too bad. If 100 messages are "open" at once, there will be bad performance either way (and probably the display code needs to be changed to only show messages that are scrolled into view anyways).

**Comment by siefkenj** [comment]
Another question: What do you think about removing `convertOutlookQuotingToBlockquote` entirely until we have a sample email to test it on? The method requires not only a DOM, but a DOM attached to a rendered tree in order to use `getComputedStyle`. This makes it very hard to test and also very hard to decouple from display code, and if we had an actual example of such a message, there'd surely be another way to find the quote text.

**Comment by siefkenj** [comment]
@Standard8 Any update on this?

**Comment by Standard8** [comment]
@siefkenj Sorry, I took a little time off from Conversations for a couple of weeks due to other priorities.

Re parsing, I think it would be an interesting idea to try out. Maybe there's an basic/rough outline or experiment could be done so it'd be easier to get an understanding of what it looks like.

For the outlook code, how about we try disabling it and pushing out an update, and seeing if it affects people? Worst case, we can always put it back in again.

**Review by Standard8 — APPROVED** [review]

**Comment by Standard8** [comment]
By the way, I've also enabled the new discussions feature now, so we can possibly move some of those discussions to there.

**Pull request merged**